Lower 1D and 2D backward convolution ops using tosa::CustomOp

pabloantoniom opened the pull request:
## Motivation

We want to add support for Conv3d lowering. However, there is no `tosa.transpose_conv3d` upstream (only `tosa.transpose_conv2D`). Therefore, we decided that the solution is to do this via tosa::CustomOp. On GPU it will be lowered to rock (easy) and on CPU it will need to be lowered to `tosa.reverse` + `tosa.conv3d`.

Because currently we are lowering both transpose Conv1d and Conv2d using `tosa.transpose_conv2D`, we want to change the current approach for 1D/2D first, so that when we introduce 3D, all transpose convs are lowered using the same path. Therefore, the current PR is in preparation of a future PR.

## Technical Details

- In `MIGraphXToTosa.cpp` we now use tosa::CustomOp to represent backward convolutions
- To lower the CustomOp representing the backward convolution:
  - CPU: A new pass called `rocmlir-custom-tosa-decompose` pass is introduced (which is based on the upstream pass).
  - GPU: In `TosaToRock.cpp`, `ConvConverter` is renamed to `ForwardConvConverter` (which now handles forward conv only), and a new `BackwardConvConverter` is added, which lowers the CustomOp representing backward convolutions. I was forced to make two separate passes because the original one was incompatible with both `Conv2dOp` and `CustomOp` ops.

## Test Plan
- Modified `migraphx-to-tosa.mlir` to check for `custom` rather than `transpose_conv2d`
- Added 2 new tests to `tosa-to-rock.mlir` to exercise backward convolution (we were missing these before...?)

## Test Result

Test passes

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

pabloantoniom committed:
Lower backwards_data_convolution to tosa.custom instead of tosa.transpose_conv2d

pabloantoniom committed:
Fix comments

pabloantoniom committed:
Add rocmlir-custom-tosa-decompose pass

pabloantoniom committed:
Fixes: Add CMake change; get attributes properly; add conversion target (otherwise it will not actually replace the op)

pabloantoniom committed:
Add TransposeConvStridedConverter

pabloantoniom committed:
Add support to lower conv1d/2d represented as tosa::CustomOp

pabloantoniom committed:
Some refactoring and making things more solid

pabloantoniom committed:
Adapt test: conv2d -> custom (+ attributes)

pabloantoniom committed:
Bring back support for forward convolution (test passed)

pabloantoniom committed:
Add bwd convolution tests (why didnt we have those...?)

pabloantoniom committed:
My bad; we dont actually want to lower conv2d (fwd) with custom op

pabloantoniom committed:
clang-format

pabloantoniom committed:
Simplify ForwardConvConverter as it should only handle fw conv

dhernandez0 commented on mlir/include/mlir/Conversion/RocMLIRPasses.td:
```diff
@@ -37,6 +37,23 @@ def ConvertRockToGPUPass : Pass<"convert-rock-to-gpu", "ModuleOp"> {
 // TosaToRock
```
RocmlirCustomTosaDecomposePass

pabloantoniom replied:
Good catch! Also changed it for the pass below

dhernandez0 commented on mlir/include/mlir/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.h:
```diff
@@ -0,0 +1,39 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass for decomposing rocmlir's custom Tosa ops
```
can you give more details here? what are the custom ops?

pabloantoniom replied:
I made it short because the other RocmlirCustom had a short header description, and I made the proper explanation in the .cpp. 

Anyway, I have added a bit more documentation here

dhernandez0 commented on mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:
is adding 3D convs a lot of more code? otherwise we could do it in this PR

pabloantoniom replied:
I think the code change *should* be minimal, but @justinrosner said he wanted to take care of that...do you mind if I do it myself Justin?

justinrosner replied:
Yes, please do go ahead!

pabloantoniom replied:
Added a separate PR to implement this here: https://github.com/ROCm/rocMLIR/pull/2034 (which is implementing the original idea of the ticket).

pabloantoniom committed:
(Partial) clang-format for td

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -688,7 +688,7 @@ static void addZeroInitPrefillAttribute(tosa::TransposeConv2DOp op,
 }
 
 template <typename OpT>
```
if it's only for forward conv, can't we use ConvOp here?

pabloantoniom replied:
It accepts Conv2DOp and Conv3DOp, which are different ops

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -728,37 +728,121 @@ class ConvConverter final : public OpConversionPattern<OpT> {
+    auto operands = adaptor.getOperands();
+    auto loc = op->getLoc();
+    auto *context = op->getContext();
+    auto input = operands[0];
```
I thought depending on the conv, input and filter might be switched?

pabloantoniom replied:
I don't know...which type of conv should have the inputs switched?

dhernandez0 replied:
backward convs

pabloantoniom replied:
I'm not sure. This function was handling backward convolution 2D before and it does not seem to be switching the input and filter. Maybe it was wrong originally?

justinrosner replied:
We have functions like this that would handle this discrepancy: https://github.com/ROCm/rocMLIR/blob/1fb11b1a9e1c8c4ae21e16dd3a799f30cb681914/mlir/lib/Dialect/Rock/IR/RockDialect.cpp#L741

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRockPass.cpp:
```diff
@@ -37,6 +40,7 @@ struct TosaToRock : public impl::TosaToRockPassBase<TosaToRock> {
   void runOnOperation() override {
     auto func = getOperation();
     if (!func->hasAttr("kernel")) {
+      LDBG() << "func has no kernel attribute, skipping pass";
```
what's the difference between this and LLVM_DEBUG(llvm::dbgs()?

pabloantoniom replied:
I didnt actually want to commit this code 😅

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -728,37 +728,121 @@ class ConvConverter final : public OpConversionPattern<OpT> {
+      return failure();
+
+    Value result;
+    Operation *rockConvOp = rockConv->getOperation();
```
same here, ConvOp?

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -728,37 +728,121 @@ class ConvConverter final : public OpConversionPattern<OpT> {
+    auto bias = operands[2];
+    RankedTensorType outputType = cast<RankedTensorType>(op.getType(0));
+
+    rock::GemmFeatures features = getGemmFeaturesFromOp(op, input.getType());
```
this code is very similar to the one in ConvConverter, can you make a common function for both?

pabloantoniom replied:
Yes, the function it's very similar, like I mentioned in the PR description, but splitting it's required because the OpConversionPattern does not like that I want to pass:

- Conv2DOp
- Conv3DOp
- CustomOp

To the same function. The problem is with ConvNDOp and CustomOp, they are not compatible.

This is a terrible hack and I'm happy to hear any suggestion

dhernandez0 replied:
is it possible to use templates to reuse code?

pabloantoniom replied:
It's possible, but it will not work. The issue is in cases like this one (there are more like this):

```
error: too few arguments to function call, single argument 'i' was not specified
  705 |     auto outputType = cast<RankedTensorType>(op.getType());
      |                                              ~~~~~~~~~~ ^
```

This compiles fine if `op` is `Conv2DOp` or `Conv3DOp`, but will not compile if it's `CustomOp`, because it expects an argument. 

The issue is that we need to call one or another depending on the templated type. This can be achieved with C++17 features like `if constexpr`, which I don't personally like. There is one single use of this idea in our codebase. I can implement it but I think it will look worse.

What I can do is refactor and reuse some code between `ForwardConvConverter` and `BackwardConvConverter`. I'll do that

dhernandez0 commented on mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:
```diff
@@ -0,0 +1,439 @@
+// mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp
+// with the added functionality of supporting transposeConv in 3D.
+// Compared to upstream, we have the following changes:
+// - We expect tosa::CustomOp instead of tosa::CustomOp
```
tosa::CustomOp instead of tosa::CustomOp ?

pabloantoniom replied:
My bad! fixed

dhernandez0 commented on mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:
```diff
@@ -0,0 +1,439 @@
+//===----------------------------------------------------------------------===//
+//
+// This pass is a downstream version of
+// mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp
```
please, can you add a commit hash to have a reference of when we copied that code?

pabloantoniom replied:
Done

dhernandez0 commented on mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:
```diff
@@ -0,0 +1,439 @@
+        rewriter.getDenseI64ArrayAttr({1, 1}),
+        /* acc_type = */ accType, /*group=*/nullptr);
+
+    rewriter.replaceOp(op, conv2d);
```
nit: replaceOpWithNewOp?

pabloantoniom replied:
This is upstream code so I'd rather leave it like it is. This file is 99% upstream code with very small variations

dhernandez0 commented on mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:
```diff
@@ -0,0 +1,439 @@
+
+  if (failed(applyPartialConversion(op, target, std::move(patterns))))
+    return signalPassFailure();
+}
```
new line

pabloantoniom replied:
Damn it, vscode 😠

pabloantoniom committed:
Address Daniel comments

pabloantoniom committed:
Delete unanted changes

pabloantoniom commented on mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:
```diff
@@ -302,8 +305,11 @@ LogicalResult ConvConverter<ConvType>::matchAndRewrite(
+      cop = tosa::CustomOp::create(
           rewriter, loc, new1DOutTy,
+          /* operator_name */ "conv_bwd_data",
+          /* domain_name  */ "rock",
```
This should be 'rocmlir' (to follow `RocmlirCustomTosaToLinalg`), not 'rock'. And also declare it somewhere rather than hardcoding it here.

pabloantoniom commented on mlir/test/Conversion/MIGraphXToTosa/migraphx-to-tosa.mlir:
```diff
@@ -209,8 +209,8 @@ func.func @quant_conv2d_float8(%arg0: !migraphx.shaped<1x16x4x4xf8E5M2, 256x16x4
 // CHECK-LABEL: @bwd_data_conv
```
`bwd_data_conv2d`, not `bwd_data_conv`

pabloantoniom committed:
Add tests for rocmlir-custom-tosa-decompose pass

pabloantoniom committed:
Address my own comments

pabloantoniom committed:
Forgot to commit this...

pabloantoniom committed:
Fix bug: Illegal custom ops are those with nv_bwd_data or conv_bwd_weight

pabloantoniom committed:
Add missing file to librockcompiler_deps.cmake

justinrosner commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -728,37 +728,121 @@ class ConvConverter final : public OpConversionPattern<OpT> {
+      return op->emitError("should have 1 result");
+
+    // Verify all required attributes are present. group is optional.
+    for (std::string attrName : {"pad", "stride", "dilation", "conv_kind"}) {
```
If we now have `conv_bwd_data` and `conv_bwd_weight` we no longer need `conv_kind`. It was initially added to distinguish between bwd_weight and bwd_data when both were using transpose.conv2d. 

We can remove all references to this.

pabloantoniom replied:
Absolutely, I forgot to do this! I have pushed a fix and now we just use the customOp operator_name instea of `conv_kind`.

justinrosner commented on mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:
```diff
@@ -0,0 +1,441 @@
+    ShapedType resultTy = cast<ShapedType>(op->getResult(0).getType());
+
+    // Translate acc_type, padding and stride attributes.
+    // TODO: Should we have all those attribute names as a constant and share
```
Should we file a case to track this if this is actually something that we want to do? Just so that this doesn't get lost as a TODO?

pabloantoniom replied:
> Should we file a case to track this if this is actually something that we want to do? Just so that this doesn't get lost as a TODO?

Added ticket: https://github.com/ROCm/rocMLIR-internal/issues/2037

justinrosner commented on mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:
It doesn't look like these either of these functions are handling input padding or dilations (both of which can be set on MIGraphX ops). This may end up needing the changes in https://github.com/ROCm/rocMLIR/pull/2007 that add support for these extra attributes.

pabloantoniom replied:
Yes, this would be place to add the changes from #2007. It *should* be straightforward as `RocmlirCustomTosaDecompose.cpp` very similar to upstream `TosaDecomposeTransposeConv.cpp` which is where the changes would go.

justinrosner reviewed (commented):
Have you done some running of the bwd_data e2e tests to verify that this works? e.g., https://github.com/ROCm/rocMLIR/blob/develop/mlir/test/fusion/pr-e2e/mixr-bwd-data-conv-dilation1-stride2.mlir

pabloantoniom committed:
Add common header for CustomOp-related attributes such as domain_name and operator_name

pabloantoniom committed:
Fixes

pabloantoniom commented:
> Have you done some running of the bwd_data e2e tests to verify that this works? e.g., https://github.com/ROCm/rocMLIR/blob/develop/mlir/test/fusion/pr-e2e/mixr-bwd-data-conv-dilation1-stride2.mlir

Yep, seems to work:

```bash
$ LIT_FILTER=mixr-bwd-data-conv-dilation1-stride2 ninja check-rocmlir
[3/228] Running utility command for fusion-e2e-tests
DONE!
[227/228] Running the RocMLIR regression tests

Testing Time: 42.91s

Total Discovered Tests: 1020
  Excluded: 1019 (99.90%)
  Passed  :    1 (0.10%)
```

pabloantoniom committed:
Delete conv_kind, we will use customOp operator_name instead

pabloantoniom committed:
Run clang-format

pabloantoniom committed:
New line...

pabloantoniom committed:
New line...

pabloantoniom committed:
Remove TODO; opened ticket to track it

pabloantoniom committed:
Fix test; after we removed conv_kind, we should now look for operator_name

pabloantoniom committed:
Refactor in ForwardConvConverter/BackwardConvConverter

dhernandez0 commented on mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:
```diff
@@ -302,8 +307,11 @@ LogicalResult ConvConverter<ConvType>::matchAndRewrite(
-      cop = tosa::TransposeConv2DOp::create(
+      cop = tosa::CustomOp::create(
           rewriter, loc, new1DOutTy,
+          /* operator_name */ ROCK_CUSTOMOP_CONV_BWD_DATA,
```
I don't think we use this style in the codebase? I think it's 
```
/*operator_name=*/ROCK_CUSTOMOP_CONV_BWD_DATA,
```

Not sure why it's not enforced by clang format.

pabloantoniom replied:
Added to https://github.com/ROCm/rocMLIR-internal/issues/2011 ticket for further investigation

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -325,7 +326,7 @@ makeRockConv(ConversionPatternRewriter &rw, Operation *op, Value input,
 
   Operation *cop = nullptr;
-  if (convKind == "bwd_data") {
+  if (convKind == ROCK_CUSTOMOP_CONV_BWD_DATA) {
```
nit: convBackwardKind? to highlight it's only supposed to be passed for backward. Also, we could make it std::optional

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -728,65 +759,112 @@ class ConvConverter final : public OpConversionPattern<OpT> {
+        op.getOperatorName() != ROCK_CUSTOMOP_CONV_BWD_WEIGHT)
+      return op->emitError("has an invalid operator_name");
+    if (op.getNumOperands() < 5)
+      return op->emitError("should have 5 or more operands");
```
nit: should -> must? same below

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -728,65 +759,112 @@ class ConvConverter final : public OpConversionPattern<OpT> {
+    if (op.getNumResults() != 1)
+      return op->emitError("should have 1 result");
+
+    // Verify all required attributes are present. group is optional.
```
nit: group -> "group"

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRockPass.cpp:
```diff
@@ -70,8 +71,13 @@ struct TosaToRock : public impl::TosaToRockPassBase<TosaToRock> {
+    target.addDynamicallyLegalOp<tosa::CustomOp>([](tosa::CustomOp op) {
+      return op.getDomainName() != ROCK_CUSTOMOP_DOMAIN_NAME ||
+             (op.getOperatorName() != ROCK_CUSTOMOP_CONV_BWD_DATA &&
+              op.getOperatorName() != ROCK_CUSTOMOP_CONV_BWD_WEIGHT);
```
I thought we didn't support backward weight?

pabloantoniom replied:
We don't, but because we have `ROCK_CUSTOMOP_CONV_BWD_WEIGHT` in some places I have it's good for completeness to have it checked here.

dhernandez0 commented on mlir/test/Conversion/MIGraphXToTosa/migraphx-to-tosa.mlir:
do we have tests with stride>1, group>1 and dilation>1?

pabloantoniom replied:
Actually the test was for stride > 1. So I have:
- Renamed conv2d test to conv2d_strided
- Added a test for normal conv2 (no stride)
- Added tests for group > 1 and dilation > 1

It's fair to have those tests, and a pity at the same time that we didnt have those before, specially because now this PR looks tremendous but it's actually not that big, it's just because we are adding a bunch of tests 😅

dhernandez0 commented on mlir/test/Conversion/RocmlirCustomTosaDecompose/rocmlir-custom-tosa-decompose.mlir:
same here, can we have cases with group, dilation and stride?

dhernandez0 commented on mlir/test/Conversion/TosaToRock/tosa-to-rock.mlir:
same, group, dilation, stride?

pabloantoniom committed:
Address nits

pabloantoniom committed:
For all tests (migraphx-to-tosa, rocmlir-custom-tosa-decompose.mlir, tosa-to-rock):

- Renamed conv2d to conv2d_stride (as it was actually strided)
- Added conv2d test (with no stride) and tests for group and dilation

dhernandez0 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -728,65 +762,112 @@ class ConvConverter final : public OpConversionPattern<OpT> {
-                                   ValueRange{result, biasExpand});
+    rock::GemmFeatures features = getGemmFeaturesFromOp(op, input.getType());
+
+    if (failed(setSplitKAttrs(op, features, rw)))
```
sorry I forgot, can we add a test for this? do we set split-k correctly?

pabloantoniom replied:
Done!

dhernandez0 commented on mlir/test/Conversion/TosaToRock/tosa-to-rockk.mlir:
fix name of the file: rockk -> rock

pabloantoniom replied:
Removed, didnt want to commit that 🤦

pabloantoniom committed:
Delete unwanted file and add split-k test

pabloantoniom committed:
Fix newline
